lib/koala/api/graph_batch_api.rb

```diff
@@ -41,67 +41,106 @@ def graph_call_in_batch(path, args = {}, verb = "get", options = {}, &post_proce
-          index = 0 # keep compat with ruby 1.8 - no with_index for map
-          response.map do |call_result|
-            # Get the options hash
-            batch_op = batch_calls[index]
-            index += 1
-
-            raw_result = nil
-            if call_result
-              parsed_headers = if call_result.has_key?('headers')
-                call_result['headers'].inject({}) { |headers, h| headers[h['name']] = h['value']; headers}
-              else
-                {}
-              end
-
-              if (error = check_response(call_result['code'], call_result['body'].to_s, parsed_headers))
-                raw_result = error
-              else
-                # (see note in regular api method about JSON parsing)
-                body = JSON.load("[#{call_result['body'].to_s}]")[0]
-
-                # Get the HTTP component they want
-                raw_result = case batch_op.http_options[:http_component]
-                when :status
-                  call_result["code"].to_i
-                when :headers
-                  # facebook returns the headers as an array of k/v pairs, but we want a regular hash
-                  parsed_headers
-                else
-                  body
-                end
-              end
-            end
-
-            # turn any results that are pageable into GraphCollections
-            # and pass to post-processing callback if given
-            result = GraphCollection.evaluate(raw_result, @original_api)
-            if batch_op.post_processing
-              batch_op.post_processing.call(result)
-            else
-              result
-            end
+      def generate_results
+        index = 0
+        lambda do |call_result|
+          batch_op     = batch_calls[index]; index += 1
+          post_process = batch_op.post_processing
+
+          # turn any results that are pageable into GraphCollections
+          result = GraphCollection.evaluate(
+            result_from_response(call_result, batch_op),
+            original_api
+          )
+
+          # and pass to post-processing callback if given
+          if post_process
+            post_process.call(result)
+          else
+            result
           end
         end
       end
 
+      def bad_response
+        # Facebook sometimes reportedly returns an empty body at times
+        BadFacebookResponse.new(200, '', 'Facebook returned an empty body')
+      end
+
+      def result_from_response(response, options)
+        return nil if response.nil?
+
+        headers   = coerced_headers_from_response(response)
+        error     = error_from_response(response, headers)
+        component = options.http_options[:http_component]
+
+        error || result_from_component({
+          :component => component,
+          :response  => response,
+          :headers   => headers
+        })
+      end
+
+      def coerced_headers_from_response(response)
+        headers = response.fetch('headers', [])
```

Comment:
This introduces a behavior change -- if somehow the field isn't present, it'll raise a `KeyError`. where previously it would just return nil for the headers (old line 68). That might ultimately be the right behavior but I'd prefer to isolate those changes from the general refactoring -- could you preserve the old behavior?